pkg/process/procutil/process_linux.go

```diff
@@ -338,6 +358,96 @@ func (p *Probe) parseStatusKV(key, value string, sInfo *statusInfo) {
 	}
 }
 
+// parseStat retrieves stat info from "stat" file for a process in procfs
+func (p *Probe) parseStat(pidPath string, pid int32, now time.Time) *statInfo {
+	path := filepath.Join(pidPath, "stat")
+	var err error
+
+	sInfo := &statInfo{
+		cpuStat: &CPUTimesStat{},
+	}
+
+	contents, err := ioutil.ReadFile(path)
+	if err != nil {
+		return sInfo
+	}
+
+	sInfo = p.parseStatContent(contents, sInfo, pid, now)
+	return sInfo
+}
+
+// parseStatContent takes the content of "stat" file and parses the values we care about
+func (p *Probe) parseStatContent(statContent []byte, sInfo *statInfo, pid int32, now time.Time) *statInfo {
+	// We want to skip past the executable name, which is wrapped in one or more parenthesis
+	startIndex := bytes.LastIndexByte(statContent, byte(')'))
+	if startIndex == -1 || startIndex+1 >= len(statContent) {
+		return sInfo
+	}
+
+	content := statContent[startIndex+1:]
+	// use spaces and prevChartIsSpace to simulate strings.Fields() to avoid allocation
+	spaces := 0
+	prevCharIsSpace := false
+	var ppidStr, utimeStr, stimeStr, startTimeStr string
+
+	for i := range content {
+		if unicode.IsSpace(rune(content[i])) {
+			if !prevCharIsSpace {
+				spaces++
+			}
+			prevCharIsSpace = true
+			continue
+		} else {
+			prevCharIsSpace = false
+		}
+
+		if spaces == 2 {
+			ppidStr += string(content[i])
+		} else if spaces == 12 {
+			utimeStr += string(content[i])
+		} else if spaces == 13 {
+			stimeStr += string(content[i])
+		} else if spaces == 20 {
+			startTimeStr += string(content[i])
+		}
+	}
+
+	if spaces <= 20 { // We access index 20 and below, so this is just a safety check.
+		return sInfo
+	}
+
+	ppid, err := strconv.ParseInt(ppidStr, 10, 32)
+	if err == nil {
+		sInfo.ppid = int32(ppid)
+	}
+
+	utime, err := strconv.ParseFloat(utimeStr, 64)
+	if err == nil {
+		sInfo.cpuStat.User = utime / ClockTicks
+	}
+	stime, err := strconv.ParseFloat(stimeStr, 64)
+	if err == nil {
+		sInfo.cpuStat.System = stime / ClockTicks
+	}
+	// the nice parameter location seems to be different for various procfs,
+	// so we fetch that using syscall
+	snice, err := syscall.Getpriority(syscall.PRIO_PROCESS, int(pid))
+	if err == nil {
+		sInfo.nice = int32(snice)
+	}
+
+	sInfo.cpuStat.CPU = "cpu"
+	sInfo.cpuStat.Timestamp = now.Unix()
+
+	t, err := strconv.ParseUint(startTimeStr, 10, 64)
+	if err == nil {
+		ctime := (t / uint64(ClockTicks)) + p.bootTime
+		sInfo.createTime = int64(ctime * 1000)
```

Comment:
Can you add a comment here why we're multiplying by 1000?

Reply:
Added. The logic is mirrored from gopsutil so we get the same format from process-agent point of view.